parsec/core/gui/trio_thread.py

```diff
@@ -200,6 +205,62 @@ def _run_job(self, job, *args, sync=False):
             logger.info(f"The submitted job `{job}` won't run as the trio loop is not running")
             raise JobSchedulerNotAvailable("The trio loop is not running")
 
+    def submit_throttled_job(
+        self, throttling_id: str, delay: float, qt_on_success, qt_on_error, fn, *args, **kwargs
+    ):
+        """
+        Throttle `fn` execution: immediatly execute the job unless it has been
+        already executed in the last `delay` seconds, in which case we schedule
+        the execution to occurs when the delay is elapsed.
+        If `fn` is already scheduled, the args/kwargs will be updated (i.e.
+        only one `fn` execution is scheduled at a time) and the schedule time
+        will be set to the sooner reaching the delay.
+        """
+        # Fool-proof sanity check, signals must be wrapped in `ThreadSafeQtSignal`
+        assert not [x for x in args if isinstance(x, pyqtBoundSignal)]
+        assert not [v for v in kwargs.values() if isinstance(v, pyqtBoundSignal)]
+
+        async def _throttled_execute(task_status=trio.TASK_STATUS_IGNORED):
+            nonlocal qt_on_success, qt_on_error, args, kwargs
+            submitted = trio.current_time()
+
+            # Only modify `_scheduled_jobs_args` from the trio
+            # thread to avoid concurrent acces with the Qt thread
+            self._scheduled_jobs_args[throttling_id] = (qt_on_success, qt_on_error, args, kwargs)
+            task_status.started()
+
+            last_executed = self._scheduled_jobs_last_executed.get(throttling_id, 0)
+            await trio.sleep_until(last_executed + delay)
```

Comment:
Maybe add a comment to show that sleeping is only needed if a similar job ran in the last `delay` seconds.

Reply:
yup, I love comments !